Make kui_tree real C++

gahr opened the pull request:
As a side effect, we can get rid of a few checks for nullptr, reduce the
number of allocations, we provide better encapsulation, and a clearer
API.

Signed-off-by: Pietro Cerutti <gahr@gahr.ch>

brasko commented:
Memory Allocation
    - kui_tree.h:kui_tree() has a comment,
      A new instance on success, or throw std::bad_alloc on error.
      Up until now, all the allocation functions are in
      lib/util/sys_util.h
      and they specify that
      If a memory allocation fails cgdb will exit
      How should we handle this in C++?

Some style guidelines
    - I still use 4 spaces for indentation (I didn't see any issues here)
    - I use 80 char terminals, and would like to not see the code or
      comments overflow
    - Some doxygen comments in the C++ classes continue to have the
      @param ktree, even though it is no longer there

Nested classes vs keeping the classes defined in the C++ code
    - For cases like kui_tree.h::node, I think i'd prefer to keep the node
      class in kui_tree.cpp instead of a nested private class.

      This would also keep memory, and map includes in the C++ file, instead
      of moving them to the header file.

      What are your thoughts here?

In the below 2 code fragments, it's surprising to me that there are no
error checks. Generally speaking, if you insert or erase, the next thing
you do is check for an error.
    diff --git a/lib/kui/kui.cpp b/lib/kui/kui.cpp
    index e6e1804..659d27e 100644
    --- a/lib/kui/kui.cpp
    +++ b/lib/kui/kui.cpp
    @@ -339,8 +328,7 @@ int kui_ms_register_map(struct kui_map_set *kui_ms,

         kui_ms->maps[key_data] = map;

    -    if (kui_tree_insert(kui_ms->ktree, map->literal_key, map) == -1)
    -        return -1;
    +    kui_ms->ktree.insert(map->literal_key, map);

         return 0;
     }
    @@ -356,8 +344,7 @@ int kui_ms_deregister_map(struct kui_map_set *kui_ms, const char *key)
         }

         /* Delete from the tree */
    -    if (kui_tree_delete(kui_ms->ktree, iter->second->literal_key) == -1)
    -        return -1;
    +    kui_ms->ktree.erase(iter->second->literal_key);

         kui_map_destroy(iter->second);
         kui_ms->maps.erase(iter);

gahr commented:
> Memory Allocation - kui_tree.h:kui_tree() has a comment, A new instance on success, or throw std::bad_alloc on error. Up until now, all the allocation functions are in lib/util/sys_util.h and they specify that If a memory allocation fails cgdb will exit How should we handle this in C++?

The [old code](https://github.com/cgdb/cgdb/blob/ea5edddb6dc4ac3f04b6981cd303e3f6ecaefc53/lib/kui/kui.cpp#L289-L293) is wrong: [kui_tree_create](https://github.com/cgdb/cgdb/blob/ea5edddb6dc4ac3f04b6981cd303e3f6ecaefc53/lib/kui/kui_tree.cpp#L160-L166) uses a `new` expression on a `kui_tree` type, which never returns `nullptr`. I guess the reason why the check is there is because the code was previously using malloc. I mentioned `std::bad_alloc` in the doxy comment for the sake of completeness, but I guess we can skip it everywhere because client code is not going to handle it anyway: if we can't allocate memory there's not much we can do, including printing any decent error message.
 
> Some style guidelines - I still use 4 spaces for indentation (I didn't see any issues here)

Agreed.

I use 80 char terminals, and would like to not see the code or comments overflow

Fixed in da47605.

> Some doxygen comments in the C++ classes continue to have the @param ktree, even though it is no longer there

Fixed in d361798.

> Nested classes vs keeping the classes defined in the C++ code - For cases like kui_tree.h::node, I think i'd prefer to keep the node class in kui_tree.cpp instead of a nested private class.  This would also keep memory, and map includes in the C++ file, instead of moving them to the header file. What are your thoughts here?

Good idea, done in c9138cb. I still need `<memory>` in the header file for `std::shared_ptr`, though.


> 
> In the below 2 code fragments, it's surprising to me that there are no error checks. Generally speaking, if you insert or erase, the next thing you do is check for an error. diff --git a/lib/kui/kui.cpp b/lib/kui/kui.cpp index e6e1804..659d27e 100644 --- a/lib/kui/kui.cpp +++ b/lib/kui/kui.cpp @@ -339,8 +328,7 @@ int kui_ms_register_map(struct kui_map_set *kui_ms,
> 
> ```
>      kui_ms->maps[key_data] = map;
> 
> -    if (kui_tree_insert(kui_ms->ktree, map->literal_key, map) == -1)
> -        return -1;
> +    kui_ms->ktree.insert(map->literal_key, map);
> 
>      return 0;
>  }
> @@ -356,8 +344,7 @@ int kui_ms_deregister_map(struct kui_map_set *kui_ms, const char *key)
>      }
> 
>      /* Delete from the tree */
> -    if (kui_tree_delete(kui_ms->ktree, iter->second->literal_key) == -1)
> -        return -1;
> +    kui_ms->ktree.erase(iter->second->literal_key);
> 
>      kui_map_destroy(iter->second);
>      kui_ms->maps.erase(iter);
> ```

Honestly, I don't see a failure case here for either `insert` or `erase`
* `erase` is the easiest to check, it [can't fail](https://github.com/cgdb/cgdb/blob/ea5edddb6dc4ac3f04b6981cd303e3f6ecaefc53/lib/kui/kui_tree.cpp#L173-L177) even on master;
* as for `insert`,  I [do check](https://github.com/cgdb/cgdb/blob/da47605c0954b833e0ccd3e56c843035de155cb5/lib/kui/kui.cpp#L319-L321) that it gets passed meaningful arguments so that it doesn't try to dereference a `nullptr` (in the future, I'd like to get rid of those bare pointers too); given valid arguments, `kui_tree::insert` will not fail: it will either insert a new node or modify the value of an existing node.

brasko commented:
You forgot to sign off on the 4 most recent patches. Please do that and i can move to integration.
